### PR TITLE
[Feature] WorldModel + WorldModelLoss — general model-based RL abstraction

### LINKED ISSUE #3774: [RFC] torchrl.modules.WorldModel — A General TensorDict-Native World Model Abstraction
## Overview

This RFC proposes adding `torchrl.modules.WorldModel`, a composable, TensorDict-native abstraction for learned environment dynamics, to serve as a general foundation for model-based RL workflows in TorchRL.

---

## Motivation

TorchRL has strong abstractions for environments (`EnvBase`), policies (`TensorDictModule`), data collection (`Collector`), replay buffers, and loss modules. Model-based RL, however, still requires users to write substantial glue code.

### What exists today

TorchRL ships Dreamer-specific components:

- `RSSMPrior`, `RSSMPosterior`, `RSSMRollout` — RSSM-specific dynamics (also V3 discrete variants)
- `ObsEncoder`, `ObsDecoder` — Conv stacks for pixel observations
- `DreamerActor` — distribution head for the imagination policy
- `WorldModelWrapper` — thin `TensorDictSequential(transition_model, reward_model)` with no rollout interface
- `ModelBasedEnvBase` / `DreamerEnv` — environment wrapper that calls `world_model(td)` in `_step()`
- `DreamerModelLoss`, `DreamerActorLoss`, `DreamerValueLoss` — losses coupled tightly to the RSSM + DreamerEnv stack

### The gap

There is no general abstraction for learned dynamics. Users implementing any world-model-based algorithm (MBPO, TD-MPC, DreamerV3, PlaNet, Iris, CWVAE, or a custom learned env) must:

1. Manually wrap each component with `TensorDictModule` and wire `in_keys`/`out_keys` by hand
2. Write a custom multi-step rollout loop (no standard `rollout(policy, horizon)` interface)
3. Accept that the losses (`DreamerActorLoss`) are tightly coupled to `DreamerEnv`, not to a generic dynamics interface
4. Write ~150 lines of boilerplate (`make_dreamer()` in `sota-implementations/dreamer/dreamer_utils.py`) before the training loop starts

The result: world-model workflows are effectively siloed from the rest of TorchRL. Imagined rollouts cannot be dropped into standard collectors, sequence replay buffers, or existing loss modules without significant adaptation.

### `WorldModel`

```python
from torchrl.modules import WorldModel

world_model = WorldModel(
    encoder=encoder,               # TensorDictModule: obs → latent
    dynamics=dynamics,             # TensorDictModule: (latent, action) → next_latent
    reward_head=reward_head,       # TensorDictModule: next_latent → reward
    done_head=done_head,           # TensorDictModule: next_latent → done (optional)
    decoder=decoder,               # TensorDictModule: latent → obs_recon (optional)
    observation_key="observation",
    action_key="action",
    latent_key=("latent", "state"),
    next_latent_key=("next", "latent", "state"),
)
```

**Key methods:**

```python
# Encode a real observation into the latent space
latent_td = world_model.encode(tensordict)

# Take one imagined step given current latent and action
next_td = world_model.step(tensordict)

# Decode a latent back to observation space (requires decoder)
recon_td = world_model.decode(tensordict)

# Run an imagined rollout for `horizon` steps with a given policy
# Returns TensorDict of shape [batch, horizon] — same layout as a real rollout
imagined_td = world_model.rollout(
    start_td=start_td,
    policy=actor,
    horizon=15,
)
```

### `WorldModelLoss`

```python
from torchrl.objectives import WorldModelLoss

loss_module = WorldModelLoss(
    world_model,
    losses=["latent", "reward", "done", "reconstruction"],
    kl_weight=1.0,
    reconstruction_weight=1.0,
)

real_batch = replay_buffer.sample(batch_size=256)
loss_td = loss_module(real_batch)
# loss_td contains: loss_latent, loss_reward, loss_done, loss_reconstruction
```

### End-to-end training sketch

```python
# World model update (on real data)
real_batch = replay_buffer.sample(batch_size=256)
loss_td = model_loss(real_batch)
(loss_td["loss_latent"] + loss_td["loss_reward"]).backward()
model_opt.step()

# Actor/critic update (on imagined rollouts)
start_td = real_batch[:, 0]
imagined_td = world_model.rollout(start_td, policy=actor, horizon=15)
actor_loss_td = actor_loss(imagined_td)
actor_loss_td["loss_objective"].backward()
actor_opt.step()
```

## Compatibility with Existing TorchRL Infrastructure

| Component | Compatibility |
|---|---|
| `TensorDictReplayBuffer` | Imagined rollouts stored directly (same TensorDict layout) |
| `SliceSampler` | Imagined trajectories are sequences — sample subsequences for recurrent training |
| `GAE` / `MultiAgentGAE` | Value targets computed from imagined rollouts |
| `ClipPPOLoss`, `SACLoss` | Consume imagined TensorDicts without modification |
| `EnvBase` specs | `WorldModel` optionally validates against `observation_spec`, `action_spec` |
| `Collector` / `SyncDataCollector` | Imagined rollout interface mirrors real rollout interface |
| `DreamerEnv` (existing) | Remains supported; `WorldModel.rollout` is an alternative that avoids the `EnvBase` overhead |

---

## Migration Path for Dreamer Users

Existing `DreamerModelLoss`, `DreamerActorLoss`, `DreamerValueLoss` are **unchanged**. Users may optionally migrate:

```python
# Before (current Dreamer setup — ~150 LOC in make_dreamer())
world_model = WorldModelWrapper(
    TensorDictSequential(rssm_prior, rssm_posterior, ...),
    reward_model,
)
dreamer_env = DreamerEnv(world_model, ...)
imagined_td = dreamer_env.rollout(15, actor)

# After (with WorldModel abstraction)
world_model = WorldModel(
    encoder=obs_encoder,
    dynamics=TensorDictSequential(rssm_prior, rssm_posterior),
    reward_head=reward_mlp,
    done_head=None,
)
imagined_td = world_model.rollout(start_td, actor, horizon=15)
```

## Scope of Implementation

**New files:**
- `torchrl/modules/world_model.py` — `WorldModel`
- `torchrl/objectives/world_model_loss.py` — `WorldModelLoss`
- `test/test_world_model.py` — unit tests
- `tutorials/sphinx-tutorials/world_model.py` — tutorial

**Modified files:**
- `torchrl/modules/__init__.py` — export `WorldModel`
- `torchrl/objectives/__init__.py` — export `WorldModelLoss`
- `docs/source/reference/modules_models.rst` — add entry
- `docs/source/reference/objectives_other.rst` — add entry

---

### PR DESCRIPTION
**Pull request by theap06**
Fixes #3774

- Adds \`WorldModel(TensorDictModuleBase)\`: a key-driven, architecture-agnostic composition layer for encoder + dynamics + reward/done/decoder heads, with \`encode\`, \`step\`, \`decode\` shortcuts and a \`rollout(start_td, policy, horizon)\` method whose \`[batch, horizon]\` output matches \`EnvBase.rollout\` — making imagined trajectories drop-in compatible with replay buffers, GAE, and loss modules.
- Adds \`WorldModelLoss(LossModule)\`: follows the standard \`_AcceptedKeys\` / \`set_keys()\` / \`forward() → TensorDict\` pattern; supports configurable sub-losses (\`reward\`, \`done\`, \`reconstruction\`, \`latent\`) with per-loss weights and distance-function choice.
- Existing Dreamer components (\`WorldModelWrapper\`, \`DreamerEnv\`, \`RSSMRollout\`, \`DreamerModelLoss/ActorLoss/ValueLoss\`) are **unchanged**.

## Motivation

TorchRL ships Dreamer-specific model-based components but no general abstraction. Users implementing MBPO, TD-MPC, PlaNet, or any custom world model must hand-wire \`TensorDictModule\` chains and write bespoke multi-step rollout loops. This adds the missing general layer.

## Test plan

- [x] \`python -m pytest test/test_world_model.py -v\` — 19 new tests covering forward, encode/step/decode shortcuts, nested keys, rollout shape and early termination, replay buffer compatibility, all four loss types, \`set_keys\`, per-loss weights, and gradient flow
- [x] \`python -c \"from torchrl.modules import WorldModel; from torchrl.objectives import WorldModelLoss; print('OK')\"\` — import smoke test
- [x] Existing dreamer-related tests (\`pytest test/ -k \"dreamer\" --ignore=test/llm\`) — 409 tests, all passing

### CODE REVIEW

**Comment by theap06** [comment]
@vmoens @elin-bdai this was my attempt at creating a World Model abstraction. Lmk if you have any feedback!

**Comment by theap06** [comment]
@vmoens I believe this issue is tied to the flaky tests from earlier. Lmk if the design seems sound to you.

File: torchrl/modules/tensordict_module/world_models.py [hunk 1]
```diff
@@ -4,7 +4,226 @@
...
 
-from tensordict.nn import TensorDictModule, TensorDictSequential
+from collections.abc import Callable
+
+import torch
+from tensordict import TensorDictBase
+from tensordict.nn import TensorDictModule, TensorDictModuleBase, TensorDictSequential
+
+
+class WorldModel(TensorDictModuleBase):
```
Comment (vmoens) [inline torchrl/modules/tensordict_module/world_models.py:11]: if a learned model can be stepped like an environment, it should probably be wrapped as a model-based env and use env.rollout(). Adding model.rollout() creates a second rollout implementation with subtly different semantics, and the PR already shows that risk: its policy update and done handling differ from EnvBase.rollout().

My take: this module can own prediction/composition; but an env should own rollout semantics.

**Review by vmoens — COMMENTED** [review]
I'm not sure why WorldModel is a not an env - or a module that's used by a model-based env to generate rollouts. That's the usual way to think about this in torchrl (see Dreamer for example). Can you elaborate on this choice>

File: torchrl/envs/model_based/world_model_env.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        ... )
+        >>> start_td = world_model.encode(obs_td)
+        >>> rollout = wm_env.rollout(max_steps=5, tensordict=start_td)
+        >>> rollout.shape
+        torch.Size([3, 5])
+    """
+
+    def __init__(
+        self,
+        world_model,
```
Comment (vmoens) [inline torchrl/envs/model_based/world_model_env.py]: do we have a loose type annotation for this?

File: test/test_world_model.py
Comment (vmoens) [inline test/test_world_model.py:1]: Let's try to keep a small number of test files, and move the tests to where the dreamer env classes are tested?

File: torchrl/envs/model_based/world_model_env.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        :class:`~torchrl.envs.EnvBase`'s outer ``step`` machinery expects
+        ``_step`` to return *root* keys and then re-wraps them under
+        ``("next", *)`` itself. Without this remap, ``EnvBase`` would not
+        find the reward / observation / done at the keys it expects and
+        would either crash or silently produce empty fields.
+        """
+        out = self.world_model(tensordict.clone(recurse=False))
+        next_td = out.get("next", default=None)
+
+        result = TensorDict({}, batch_size=self.batch_size, device=self.device)
```
Comment (vmoens) [inline torchrl/envs/model_based/world_model_env.py]: The `{}` is not necessary

File: torchrl/envs/model_based/world_model_env.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+        :class:`WorldModel` writes its outputs under ``("next", *)`` keys
+        (matching the layout of real env step outputs), but
+        :class:`~torchrl.envs.EnvBase`'s outer ``step`` machinery expects
+        ``_step`` to return *root* keys and then re-wraps them under
+        ``("next", *)`` itself. Without this remap, ``EnvBase`` would not
+        find the reward / observation / done at the keys it expects and
+        would either crash or silently produce empty fields.
+        """
+        out = self.world_model(tensordict.clone(recurse=False))
```
Comment (vmoens) [inline torchrl/envs/model_based/world_model_env.py]: You can use copy()

File: torchrl/envs/model_based/world_model_env.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+            result.set(key, value)
+        for key in self.full_done_spec.keys():
+            value = next_td.get(key, default=None) if next_td is not None else None
+            if value is None:
+                value = out.get(key, default=None)
+            if value is None:
+                value = torch.zeros(
+                    *self.batch_size, 1, dtype=torch.bool, device=self.device
+                )
+            result.set(key, value)
```
Comment (vmoens) [inline torchrl/envs/model_based/world_model_env.py]: should we iterate over the leaves only and include the nested specs?

File: torchrl/envs/model_based/world_model_env.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+        # determine termination on subsequent steps via ``_step``.
+        batch_shape = self.batch_size
+        out.set(
+            "done",
+            torch.zeros(*batch_shape, 1, dtype=torch.bool, device=self.device),
+        )
+        out.set(
+            "terminated",
+            torch.zeros(*batch_shape, 1, dtype=torch.bool, device=self.device),
+        )
```
Comment (vmoens) [inline torchrl/envs/model_based/world_model_env.py]: why not `spec.zeros()`?

File: torchrl/objectives/world_model_loss.py [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+
+        if "latent" in self.losses:
+            pred_latent = tensordict.get(self.tensor_keys.predicted_latent)
+            target_latent = tensordict.get(self.tensor_keys.target_latent)
+            loss_lat = distance_loss(
+                pred_latent, target_latent, self.latent_loss
+            ).mean()
+            out["loss_latent"] = self.latent_weight * loss_lat
+
+        td_out = TensorDict(out, batch_size=[])
```
Comment (vmoens) [inline torchrl/objectives/world_model_loss.py]: batch_size isn't used here, remove it

**Review by vmoens — COMMENTED** [review]
Just some minor comments but otherwise LGTM!

File: torchrl/envs/model_based/world_model_env.py (hunk 2 above)
Comment (theap06) [reply torchrl/envs/model_based/world_model_env.py]: yes, thx for the feedback

File: torchrl/envs/model_based/world_model_env.py (hunk 5 above)
Comment (theap06) [reply torchrl/envs/model_based/world_model_env.py]: the three loops now use keys(include_nested=True, leaves_only=True) over observation_spec / full_reward_spec / full_done_spec, so nested specs survive end-to-end.

**Review by vmoens — APPROVED** [review]
LGTM
do you think we should consider subclassing the dreamer classes with these for a more homogeneous implementation?

**Comment by theap06** [comment]
> LGTM do you think we should consider subclassing the dreamer classes with these for a more homogeneous implementation?

I can open an issue with the refactors for the existing dreamer classes.
